File: java/CombinationSumII.java
/*
 * Filename: CombinationSumII.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Wed 08 Apr 2015 11:38:45 PM EDT
 */

public class Solution {
	public List<List<Integer>> combinationSum2(int[] num, int target) {
		List<List<Integer>> ret = new ArrayList<List<Integer>>();
		List<Integer> tuple = new ArrayList<Integer>();

		if (num == null || num.length == 0) {
			return ret;
		}

		Arrays.sort(num);

		combinationSum2Aux(num, target, tuple, ret, 0);
		return ret;
	}

	private void combinationSum2Aux(int[] num, int target, List<Integer> tuple, List<List<Integer>> ret, int bound) {
		if (target == 0) {
			ret.add(new ArrayList<Integer>(tuple));
			return;
		}

		if (target < 0) {
			return;
		}

		for (int i = bound; i < num.length; i++) {
			if (i != bound && num[i] == num[i-1]) {
				continue;
			}
			tuple.add(num[i]);
			combinationSum2Aux(num, target - num[i], tuple, ret, i+1);
			tuple.remove(tuple.size() - 1);
		}
	}
}

File: cpp/combination_sum_ii.cc
/*
 * Filename: combination_sum_ii.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sat 25 Jul 2015 08:42:53 PM EDT
 */

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
		vector<vector<int> > combinationSum2(vector<int>& candidates, int target) {
			vector<vector<int> > res;
			vector<int> v;
			sort(candidates.begin(), candidates.end());
			combination_sum2_aux(candidates, target, -1, res, v);
			return res;
		}

	private:
		void combination_sum2_aux(vector<int>& candidates, int target, int i, vector<vector<int> >& res, vector<int>& v) {
			if (target == 0) {
				res.push_back(v);
			} else if (target > 0) {
				for (vector<int>::size_type j = i+1; j < candidates.size(); j++) {
					if (j-i-1 && candidates[j] == candidates[j-1]) {
						continue;
					}

					v.push_back(candidates[j]);
					combination_sum2_aux(candidates, target - candidates[j], j, res, v);
					v.pop_back();
				}
			}
		}
};
